[Detection Engine] Cypress - Add more robust selection from our DataView dropdown component

rylnd opened the pull request:
This addresses some recent cypress failures: 

* https://github.com/elastic/kibana/issues/212743 (Rule Creation with a DataView)
* https://github.com/elastic/kibana/issues/212742 (Rule Creation + Edit with a DataView)
* https://github.com/elastic/kibana/issues/213752 (Rule Creation + Filter with a DataView)

This appears (as much as a cypress failure can 😓) to be caused by an incorrect/false-positive assertion, leading to us (very occasionally) interacting with the combobox before it's ready. We were calling `.should('not.be.disabled')` on an element that could never be disabled. By calling that instead on the inner `input` that actually is enabled/disabled, we have the sanity check that was originally intended.

This PR also adds a post-action check (`.should('contains', thingThatWasTyped)`) so that if the action fails, the test doesn't fail inscrutably at a later step.

### Evidence

* https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8003 (50x)
* https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8004 (200x)
* https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8033 (200x)

### Significance

**Note also** that some initial investigation found this pattern in several places in our test suite. I'm going to follow up on this focused PR with a more comprehensive one (once this is proven out in the flaky runner).

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed




<!--ONMERGE {"backportTargets":["8.16","8.17","8.18","8.x"]} ONMERGE-->

rylnd committed:
Add more robust selection from our dropdown component

This adds a notable "downArrow" to the way that we select our dataview
from the dropdown. In recent work on the related form UI for alert
suppression, I found that the `.type(`thing{enter}`) only works if
"thing" is the only option in the dropdown; if there are more than one,
you need to explicitly "select" the correct item and then hit enter.

My theory is that something about how those dataviews are populated has
changed: either another (earlier) test has started creating dataviews,
or this is a consequence of the default alerts dataview now showing up
here. If this is the case, then this should fix these tests. However,
cypress is crashing about 99% of the time locally with Chrome, so we're
gonna send this off to the flaky runner to see how it behaves.

I also added some sanity checks before/after the failing step, so that
we'll have better failure messages in case this is not the fix.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

rylnd commented:
Flaky Runner suite: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/7993

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#7993](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/7993)
[❌] [Serverless] Security Solution Detection Engine - Cypress: 74/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F213510%2Fhead)

rylnd committed:
Better sanity check around combobox interaction

The selector `[data-test-subj="comboBoxInput"]` is not actually an
`input`, and thus is never actually disabled, so this check wasn't
really doing anything.

In my investigation here, I believe that I see a period where the input
is disabled, which may be why we occasionally don't select the correct
item.

Running another round of tests on that assumption.

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#8003](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8003)
[✅] [Serverless] Security Solution Detection Engine - Cypress: 50/50 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F213510%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#8004](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8004)
[✅] [Serverless] Security Solution Detection Engine - Cypress: 200/200 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F213510%2Fhead)

rylnd commented:
That looks to have done it. I'm going to keep this PR small and apply these changes across our test suite in a follow-up.

rylnd committed:
Merge branch 'main' into rylnd/debug_cypress_dataview_failures

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/282445)
* Commit: 4e0accd352573bc115c8858463701b777a255a3c

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/282445","number":282445,"commit":"4e0accd352573bc115c8858463701b777a255a3c"}],"number":282445}
buildkite-pr-comment-->

rylnd commented:
Update: https://github.com/elastic/kibana/issues/213752 just failed this morning, and should also be fixed by this PR.

rylnd commented:
@MadameSheema looks like @elastic/security-engineering-productivity is the only codeowner here (?) but no one was assigned for review. There's not much here; this is more of an FYI.

I plan to address this interaction with EuiComboBox holistically in a followup PR, but in researching this issue I found several related issues. Even some that have been already "[fixed](https://github.com/elastic/kibana/pull/187173)" I suspect might ultimately be caused by the issue fixed here; time will tell.

MadameSheema reviewed (approved):
Thanks for fixing this flakiness!! Great work :)

rylnd commented on x-pack/test/security_solution_cypress/cypress/tasks/create_new_rule.ts:
```diff
@@ -463,7 +464,9 @@ export const removeAlertsIndex = () => {
-    cy.get(DATA_VIEW_COMBO_BOX).type(`${rule.data_view_id}{enter}`);
+    cy.get(DATA_VIEW_COMBO_BOX_INPUT).should('not.be.disabled');
+    cy.get(DATA_VIEW_COMBO_BOX).type(`${rule.data_view_id}{downArrow}{enter}`);
+    cy.get(DATA_VIEW_COMBO_BOX).should('contain', rule.data_view_id);
```
As I understand it, `have.text` will fail if the second argument is only a subset of the selected text (i.e. must be equal), whereas `contain` will fail if the second argument is absent from the selected text.

The logic represented in the above test is as follows:

1. Verify that combobox field is ready
2. Interact with combobox field
3. Assert that combobox field now represents the interaction from 2.

As I was hoping to generalize this logic and apply it to other similar (and similarly failing) tests, using `have.text` would complicate the filling of more than 1 item in a field, since it would require knowing not only the current item being filled, but also all the previous items that are in the combobox. 

Imagining future failure conditions (e.g. the EUI component changing, or the cypress matchers changing): if the assertion becomes _too loose_ and finds false positives, then the test would (confusingly, but not catastrophically) fail at a later point (as it is currently doing). If the assertion becomes _too strict_ and finds false negatives, then the test will start to fail closer to the problem.

@MadameSheema can you please explain why a more restrictive assertion is better, here? I would like to understand your reasoning before coming to a decision. Thanks!

MadameSheema replied:
Hey @rylnd!! I understand your point and is completely valid :) 

When I suggested `have.text` was because I had in mind the following scenario:

1. You have more than one data view with similar naming, i.e.: `auditbeat-exceptions-1` and `auditbeat`
2. You type `auditbeat`

If you wanted the dataview to be selected `auditbeat` but actually for wathever reason is`auditbeat-excpetions-1` with contains the test is going to pass and is not going to fail.

rylnd replied:
@MadameSheema that's a great point; I have actually encountered situations where I had to use a different field precisely because the one I was filling was a subset of another field's name 👍 .

Perhaps I can modify the selector of that assertion to look at all individual options that were filled, instead of the whole component itself; that should be the best of both worlds here.

rylnd replied:
@MadameSheema okay, I applied `have.text` everywhere that was possible in https://github.com/elastic/kibana/pull/213510/commits/05b8c69ad3e9203f4fa69712f27b330c2daf10c0.

rylnd committed:
Merge branch 'main' into rylnd/debug_cypress_dataview_failures

rylnd committed:
Defines a robust, reusable fillComboBox helper

Features:
* Asserts that the inner input is not disabled before interaction
* Asserts that each item is successfully selected

This applies the new helper to the existing "data view rule" tests, and
refactors some existing selectors (who were only used for this case) to
work with the helper.

rylnd committed:
Unskips fixed flaky tests

Note that most of this commit's diff is just whitespace.

rylnd commented:
One more flaky run to validate our latest changes: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8033

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#8033](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/8033)
[❌] [Serverless] Security Solution Detection Engine - Cypress: 174/200 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F213510%2Fhead)

marshallmain reviewed (approved):
🚀

kibanamachine commented:
Starting backport for target branches: 8.16, 8.17, 8.18, 8.x

https://github.com/elastic/kibana/actions/runs/13860697396

kibanamachine commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.16|Backport failed because of merge conflicts|
|❌|8.17|Backport failed because of merge conflicts|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/214605">](https://github.com/elastic/kibana/pull/214605)|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/214606">](https://github.com/elastic/kibana/pull/214606)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 213510
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

rylnd commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/214664">](https://github.com/elastic/kibana/pull/214664)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
